### ISSUE TITLE
Neutronium leaks ChromiumFxProcess when using DataTemplate and those are in a ItemSource in a tab control

### ISSUE THREAD

**Issue by jmecosta** [David-Desmaisons/Neutronium#6]
hi again,

notice that if the HTMLViewControl is placed inside a tab in a tab control, the ui locks up when we transition between a tab that does not contain  HTMLViewControl to a the tab that contains the HTMLViewControl.

Is this is a know issue? or something that can be tweeked from my side?

thanks in advance

**Comment by jmecosta** [comment]
i dont know if this is related however for each tab change a new process of ChromiumRederFxProcess.exe is created.

**Comment by David-Desmaisons** [comment]
1) There will be one ChromiumRederFxProcess process by html element, this is the way Chromium works.

2) Regarding the transition time, it is hapening only the first time that the element is displayed or at each transition?

**Comment by jmecosta** [comment]
![image](https://cloud.githubusercontent.com/assets/1041352/20178330/4c339994-a759-11e6-9a0f-57354b40b2e5.png)

this is after swtiching the times a couple of times, between 2 tabs. there always at anygive moment only one html element (at least is what i would want)

but i understand the tab control actually destroys the visual tree, so its expected that after each tab that contains a html  element is open a new process is created.

However it seems here, that after the that that contains the html element is closed. The process is not distroyd. So it seems its leaking processes.

The time is always the same  (about 5 secs) for each transtion to a tab that contains an html element.


Is it possible to bind the Neutronium:HTMLViewControl to a object in my view model so that it does not get created everytime?

**Comment by David-Desmaisons** [comment]
First point, 
* I just checked CEF will create two processes by component.
* These processes will only shutdown when you will dispose the corresponding neutronium component: typically this should happen when closing the corresponding component and not when the component is just hidden.

Considering your scenario:
-I assume that you have two HTMLViewControld in two tabs correct?
-Every time uou change the view model, it will have a performance impact because neutronium needs to introspect the vm to create the binding. It is a good idea, when possible, to bind once to the viewmodel.
-Other performance tips: 
* only public property of the view model are converted into javascript, so if you do need one property in the html side, you can declare it as internal
* converting command can be costly performance wise, try to design your vm in a way you minimise the number of command, see [Example.ChromiumFX.Vue.UI](https://github.com/David-Desmaisons/Neutronium/tree/master/Examples/Example.ChromiumFX.Vue.UI) for example where one command is responsable of removing various hildren elements.

**Comment by jmecosta** [comment]
in fact i you are correct, when application starts i see 2 FX processes. However i only have one HTMLViewControld in all my application. The FX proceses increase by one everytime change the tabs.

i wouldnt like to dispose the HTMLViewControld  component since that is what makes the load take timoe in first place. 

I think i can prepare a sample with the problem.

**Comment by David-Desmaisons** [comment]
Sample with the problem will be great for sure!
I will wait for it before further investigating

**Comment by jmecosta** [comment]
basic sample i cannot reproduce also, i still try to it like in the original application:

the only  think i can think of is because my tabs are dynamically created and my viewmodels are never destroyed only the views. for some reason the process is not being disposed?

```
            <TabControl controls:TabControlHelper.IsUnderlined="True" Grid.Row="1" x:Name="TabControlViews"
                        ItemsSource="{Binding BuildAllViews}" SelectedItem="{Binding SelectedView}"
                        TabStripPlacement="Bottom">
                <TabControl.Resources>
                    <Style TargetType="TabItem" BasedOn="{StaticResource MetroTabItem}">
                        <Setter Property="controls:ControlsHelper.HeaderFontSize" Value="12" />
                        <Setter Property="controls:TabControlHelper.IsUnderlined" Value="True" />
                    </Style>
                </TabControl.Resources>
                <TabControl.ItemTemplate>
                    <DataTemplate>
                        <TextBlock Text="{Binding Header}" />
                    </DataTemplate>
                </TabControl.ItemTemplate>

                <TabControl.ContentTemplate>
                    <DataTemplate>
                        <ContentControl Content="{Binding ElementName=TabControlViews, Path=SelectedItem}" />
                    </DataTemplate>
                </TabControl.ContentTemplate>
            </TabControl>
```

the normal i see while switching tabs, is that before changing there are 2 process, then i see when moving to the tab that has the html another process is created but the other one that was there is immediatly killed, and we are back to 2 processes.

So in my scnenario the killing not happening

**Comment by David-Desmaisons** [comment]
You should find a hook when the view are destroyed and then call dispose on the HTMLViewControl .

Performance-wise it will much better to have a static component loaded only once and only changing the hide-show.

**Comment by jmecosta** [comment]
ok ive made a small sample illustrating the problem of the dangling processes. 
[Neutronium Vue Application2.zip](https://github.com/David-Desmaisons/Neutronium/files/584176/Neutronium.Vue.Application2.zip)

I am not totally sure after all if the views are destroyd, in the sample ive put a destructor in the usercontrol and its not getting called. So in principle i think those views are created only once.

the ui refersh in 5 seconds is not visible in this sample, but i think its because ive reuse a existing view model and its not really small (would be nice to have attribute set on the properties that are suposed to be exported to Vue, that would solve my problem and would allow viewmodels to be shared between xaml and js since xaml performance is not impacted by the size of my model). If thats not possible, i will refactor the app to reduce the ammout of props. Even the InspectVm does not work with my viewmodel :)

**Comment by David-Desmaisons** [comment]
1) Regarding the leak, really it is linked to WPF and not to Neutronium: you can use the unload event in the Neutronium tab controler

```C#
  public BuildView()
        {
            GC.Collect();
            InitializeComponent();
            this.Unloaded += BuildView_Unloaded;
        }

        private void BuildView_Unloaded(object sender, RoutedEventArgs e) {
            this.HtmlView.Dispose();
        }
```

this will close the render process as you change the tab.


2) Performance wise, you should change your wpf model to create only once your binding to the view model.

3) The only thing you have to do to improve performance is to set internal property that are not supposed to be converted to javascript. This should not prevent you to reuse same viewmodel beetween xaml and html

**Comment by David-Desmaisons** [comment]
If you have a profiler, I could also investigate potencial performance improvements.

**Comment by David-Desmaisons** [comment]
As you updated the issue name I am closing this, as you can use the fix I provided. As mentioned it is linked to WPF component lifecycle and not tom Neutronium. Feel free to open a seperate issue for performance improvement,

**Comment by jmecosta** [comment]
yep, ive fixed the performance issue. ive moved every thing inside a contentcontrol with a new viewmodel only for the web compoennt.

Then ive used your unload event and voila works like expected. 

Thanks a lot for the help, this thread is a really good source of information.
